### DISCUSSION TITLE
my views of this language

### DISCUSSION THREAD

**Discussion by soldate** [Neat-Lang/neat#31] [General]
Hi, I'm a programmer who would like a language:
1) Simple and fast as C. "looking the c code i can see the assembler" - Linus
2) With some extra features at no/low cost, such as c++.
3) With a huge Object Oriented standard library like Java.
4) Without garbage collection.

So, i really like Neat!

but...
I really hate "?" style error propagation. 
I dont see any problem in something like:
int i = foo();
if (i == -1) { print("error!"); return -1;}
or (if you really really need "-1" as possible value)
(int i, error e) = foo();
or use try and catch.. (with "throws" but without specifying all exceptions in function declarations, like swift)

I really prefer use "?" as ternary If. All c programmers too.
i really dont see any advantage to use SumTypes, the code became too ugly i think.
i prefer instanceOf as a keyword like java, but ok.

"(int | fail FileNotFound) foo() { return "test".readAll?.itoa; }"
very ugly!

i believe that neat has a potential to kill c, c++ and java. Like D try and fail.
to achieve this, the language should be fast and easily understandable for c, c++ and java programmers.

Go Neat!

**Comment by FeepingCreature** [comment]
Yeah I agree error return typing is pretty ugly, but I really don't want to mess with exceptions. `?` is entirely there for the needs of the compiler; if you replaced every `?` in expressions like `symbol.compile(context)?.beType(__RANGE__)?` with `if (err) return err;` you would easily double the line count of the compiler.

I've been thinking of finding another syntax for `(int | fail FileNotFound) foo()`. Maybe `int foo() fail FileNotFound {}` or something? I'm really not sure there.

**Reply by FeepingCreature** [reply]
The premise behind `?` is getting into a general mindspace of "get rid of the error branches with minimum fuss". Ie. `read(fd, buf.ptr, buf.length).case(-1: return ErrnoError());` and such. The idea is to be able to see "left side of line - function invocation; right side of line - error transformation." That's the premise at least.

**Comment by soldate** [comment]
- "The idea is to be able to see "left side of line - function invocation; right side of line - error transformation."
- "you would easily double the line count of the compiler."

In my opinion, I really dont matter of writing more code. I prefer write 5-10 lines instead of 1 if the plus 5 or 10 lines made the language simpler, its ok to me. 

I think the principle order for a language should be:
Simple (readable, maintainable, "easy to see the assembler"), fast, fun and (if it doesn't disturb the previous ones) safe.

thats why c still exists today. people that create new programming languages put more and more stuff and creates a lot of complexity.

the programmers who will made your language a success are not smart like you. your language must be attractive to less intelligent people (like me :-D) and, at the same time, be capable of create really fast code for really good programmers. get their respect.
thats why GC is a mistake i guess.

a new programming language c-like just with modules (without header files) and with fat pointers could kill c, but no one create only this! 
always more and more stuff.

i think today a new "c with classes", packages, modules, Uniform Function Call Syntax, immutable by default, slice, tuple, list, dictionary, assert and a not so hard memory management is more than enough.

i dont like so much lambda or templates.. 
i really dont like rust like error handling ?, result, optionals, etc.

I think a way to convert c code to neat and neat code to c is much more important than add more features.

**Reply by FeepingCreature** [reply]
I mean, nobody forces you to use `?` :) It's a convenience feature. I'll say that I personally am absolutely not in camp "rather write 5-10 lines". That's why neat has a bunch of syntax macros, after all. If the language syntax can save me effort, I'll gladly let it.

**Comment by soldate** [comment]
Convenience feature that will be used all the time in standard library and in practically force you to use too.

"no macros! yeahhhh!" - ziglang
"a bunch of macros! yeahhhh!" - neat lang

Someone is doing it wrong! :-D

When you choose RC over GC, thinking yourself that GC is better, ITS A GREAT DECISION, because you are choosing what people want not what you want. 

KEEP THIS PRINCIPLE of doing what c-like programmers (c/c++/java/dlang millions of programmers) really wants (only the common sense, the minimum) or you will fail exactly like Walter! Each new feature, each new great way of doing the same things means a bunch of less interested programmers.

Succeeding is principle number one, not nerdy fun.

Go Neat!

**Reply by FeepingCreature** [reply]
I mean, honestly, I also need to make it a language that I want to use. If I don't enjoy my own language, I'll have failed at the most basic level; whereas if I think the language is good, I'll always have at least one fan. So yeah, it's probably not gonna get much bigger than it is now, conceptually, but stuff like macros and cute syntax is not going away. Sorry.
